# 先週の 「分類タスク実験についてのテスト」の解説

## 準備

In [1]:
### データセットの読み込み

from sklearn.datasets import load_wine
wine = load_wine()

import pandas as pd
df = pd.DataFrame(wine.data, columns = wine.feature_names)
df["class"] = wine.target

x = df.drop(["class"], axis=1)  # 説明変数 (class以外の列からなるデータフレーム)
y = df["class"]                 # 目的変数 (class列のみからなるシリーズ (データフレームの1列分))

### 訓練データとテストデータに分割
 
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.7, test_size=0.3, random_state=0)

## ランダムフォレストを用いた分類器の生成，評価

In [2]:
### 分類器の生成
print("デフォルト (n_estimators=100) の場合:")

from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=0)       # 決定木数(n_estimators)のデフォルト値は100
rf_model.fit(x_train, y_train)
rf_test = rf_model.predict(x_test)

### 評価

print("　　正解率(train): ", round(rf_model.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(rf_model.score(x_test, y_test)*100), "%")

デフォルト (n_estimators=100) の場合:
　　正解率(train):  100 %
　　正解率(test):   98 %


## FNNを用いた分類器の生成，評価

In [3]:
### 分類器の生成
print("max_iter=1000 (hidden_layer_sizes=(100,)) の場合:")

from sklearn.neural_network import MLPClassifier
mlp_model = MLPClassifier(random_state=0, max_iter=1000)    # 隠れ層のユニット数(hidden_layer_sizes)のデフォルト値は(100,)
mlp_model.fit(x_train, y_train)
rf_test = mlp_model.predict(x_test)

### 評価

print("　　正解率(train): ", round(mlp_model.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(mlp_model.score(x_test, y_test)*100), "%")

max_iter=1000 (hidden_layer_sizes=(100,)) の場合:
　　正解率(train):  100 %
　　正解率(test):   96 %


- 訓練データでの正解率は，共に <u>100%</u> だったが，テストデータでの正解率は，ランダムフォレストでは
  <u>98%</u> だったのに対して，FNN では <u>96%</u> だったので，正解率の高い**ランダムフォレストの方が高精度**だったといえる

## ランダムフォレストの修正

In [4]:
### n_estimators=10 の場合
print("n_estimators=10 の場合:")

rf_model1 = RandomForestClassifier(random_state=0, n_estimators=10)
rf_model1.fit(x_train, y_train)
rf_test = rf_model1.predict(x_test)

### 評価

print("　　正解率(train): ", round(rf_model1.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(rf_model1.score(x_test, y_test)*100), "%")

n_estimators=10 の場合:
　　正解率(train):  100 %
　　正解率(test):   100 %


In [5]:
### n_estimators=1 の場合
print("n_estimators=1 の場合:")

rf_model2 = RandomForestClassifier(random_state=0, n_estimators=1)
rf_model2.fit(x_train, y_train)
rf_test = rf_model2.predict(x_test)

### 評価

print("　　正解率(train): ", round(rf_model2.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(rf_model2.score(x_test, y_test)*100), "%")

n_estimators=1 の場合:
　　正解率(train):  94 %
　　正解率(test):   85 %


## FNNの修正

In [6]:
### hidden_layer_sizes=(50,) の場合
print("max_iter=1000, hidden_layer_sizes=(50,) の場合:")

mlp_model1 = MLPClassifier(random_state=0, max_iter=1000, hidden_layer_sizes=(50,))
mlp_model1.fit(x_train, y_train)
mlp_test = mlp_model1.predict(x_test)

### 評価

print("　　正解率(train): ", round(mlp_model1.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(mlp_model1.score(x_test, y_test)*100), "%")

max_iter=1000, hidden_layer_sizes=(50,) の場合:
　　正解率(train):  19 %
　　正解率(test):   22 %


In [7]:
### hidden_layer_sizes=(400,) の場合
print("max_iter=1000, hidden_layer_sizes=(400,) の場合:")

mlp_model2 = MLPClassifier(random_state=0, max_iter=1000, hidden_layer_sizes=(400,))
mlp_model2.fit(x_train, y_train)
mlp_test = mlp_model2.predict(x_test)

### 評価

print("　　正解率(train): ", round(mlp_model2.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(mlp_model2.score(x_test, y_test)*100), "%")

max_iter=1000, hidden_layer_sizes=(400,) の場合:
　　正解率(train):  99 %
　　正解率(test):   94 %


In [8]:
### hidden_layer_sizes=(400,), max_iter=200 の場合
print("max_iter=200, hidden_layer_sizes=(400,) の場合:")

mlp_model3 = MLPClassifier(random_state=0, max_iter=200, hidden_layer_sizes=(400,))
mlp_model3.fit(x_train, y_train)
mlp_test = mlp_model3.predict(x_test)

### 評価

print("　　正解率(train): ", round(mlp_model3.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(mlp_model3.score(x_test, y_test)*100), "%")

max_iter=200, hidden_layer_sizes=(400,) の場合:
　　正解率(train):  96 %
　　正解率(test):   98 %


c:\ProgramData\Miniconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [9]:
### hidden_layer_sizes=(100,), max_iter=200 の場合
print("max_iter=200, hidden_layer_sizes=(100,) の場合:")

mlp_model4 = MLPClassifier(random_state=0, max_iter=200, hidden_layer_sizes=(100,))
mlp_model4.fit(x_train, y_train)
mlp_test = mlp_model4.predict(x_test)

### 評価

print("　　正解率(train): ", round(mlp_model4.score(x_train, y_train)*100), "%")
print("　　正解率(test):  ", round(mlp_model4.score(x_test, y_test)*100), "%")

max_iter=200, hidden_layer_sizes=(100,) の場合:
　　正解率(train):  91 %
　　正解率(test):   93 %


c:\ProgramData\Miniconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## テストデータに対する正解率まとめ

In [11]:
print("### ランダムフォレスト\n")

print("　　決定木数100 (default):          ", round(rf_model.score(x_test, y_test)*100), "%")
print("　　決定木数 10 (n_estimators=10): ", round(rf_model1.score(x_train, y_train)*100), "%")
print("　　決定木数  1 (n_estimators=1):   ", round(rf_model2.score(x_train, y_train)*100), "%")

print("\n### FNN，収束時 (max_iter=1000)\n")

print("　　隠れ層ユニット数 50 (hidden_layer_sizes=(50,)): ", round(mlp_model1.score(x_test, y_test)*100), "%")
print("　　隠れ層ユニット数100 (default):                  ", round(mlp_model.score(x_test, y_test)*100), "%")
print("　　隠れ層ユニット数400 (hidden_layer_sizes=(400,)):", round(mlp_model2.score(x_test, y_test)*100), "%")

print("\n### FNN，学習反復回数200回 (max_iter=200)\n")

print("　　隠れ層ユニット数100 (hidden_layer_sizes=(100,)):", round(mlp_model4.score(x_test, y_test)*100), "%")
print("　　隠れ層ユニット数400 (hidden_layer_sizes=(400,)):", round(mlp_model3.score(x_test, y_test)*100), "%")




### ランダムフォレスト

　　決定木数100 (default):           98 %
　　決定木数 10 (n_estimators=10):  100 %
　　決定木数  1 (n_estimators=1):    94 %

### FNN，収束時 (max_iter=1000)

　　隠れ層ユニット数 50 (hidden_layer_sizes=(50,)):  22 %
　　隠れ層ユニット数100 (default):                   96 %
　　隠れ層ユニット数400 (hidden_layer_sizes=(400,)): 94 %

### FNN，学習反復回数200回 (max_iter=200)

　　隠れ層ユニット数100 (hidden_layer_sizes=(100,)): 93 %
　　隠れ層ユニット数400 (hidden_layer_sizes=(400,)): 98 %


- この例におけるランダムフォレストによる分類器においては，決定木の数が多いほどテストデータに対する精度が高い？

  決定木を指定しなかったとき (決定木数100のとき) の正解率 (98%) よりも，決定木数10の時の正解率
  (100%) の方が高い．したがって **×**

- この例におけるFNNによる分類器においては，隠れ層のユニット数が多いほどテストデータに対する精度が高い？

  ユニット数を指定しなかったとき (ユニット数100のとき) の正解率 (96%) の方が，決定木数400の時の正解率
  (94%) よりも高い．したがって **×**

- この例におけるFNNによる分類器において，学習の反復を収束前に打ち切ると，テストデータに対する精度がは必ず低下する？

  ユニット数が400の場合，反復回数を収束前の200回で打ち切った場合の正解率 (98%) の方が，収束するまで反復した場合
  (max_iter=1000) の正解率 (94%) よりも高い．したがって **×**

#### 元ファイル
- [classify.ipynb](classify.ipynb)